allennlp/data/dataset_readers/dataset_reader.py

```diff
@@ -304,143 +198,135 @@ def text_to_instance(self, *inputs) -> Instance:
         """
         raise NotImplementedError
 
-    def serialize_instance(self, instance: Instance) -> str:
+    def apply_token_indexers(self, instance: Instance) -> None:
```

Comment:
I'm pretty sure we can implement this for you, both to remove them and to restore them.  It's trivial if the field is just a `TextField` (store a map from field name to dictionary, using `id()` to make sure it's the same object per field name and crashing if it's not).  The only other case to handle is `ListField[TextField]`, which also shouldn't be hard, you just need to store how many nested lists there are.  This could be done in the base class implementation, allowing people to override it if they have some crazy new field type that they implement, but the default should just work, to both remove and restore the indexers, so no one has to implement anything.

Reply:
Hmm the problem I see with this is that the "removing token indexers" and "restoring the token indexers" steps happen in separate processes. The "removing" step in a worker and the "restoring" step in the main process. So without some multi-process communication, only the worker process would know what the mapping is.

I'll put some more thought into this though.

Reply:
Hmm, ok.  Another potential solution, again as a default implementation that would work for most use cases, and can be overridden for special cases: just assume that there's a `self._token_indexers`, which is the only one that gets used.  Then you don't need any communication between processes.

Reply:
That seems reasonable. And we can also check to make sure that `self._token_indexers` is the only `TokenIndexer` attribute of the `DatasetReader`, if it exists. Throwing an error if we find other token indexers.

Reply:
On second thought, I'm not sure this is the right way to go.

My main concern is that this default implementation of `apply_token_indexers` would not be very efficient since there's no guarantee what fields an `Instance` can contain, and therefore `apply_token_indexers` would have to inspect each `Instance` to some degree.

Reply:
Yeah, but the difference between train and test will typically not be in `TextFields`.  It's probably sufficient to just guess what fields need to be modified from the first `Instance`, log what the guess is, and proceed.  If something crashes, catch the error and give a nicer message saying that our guess was wrong, and you need to override this method for multi-processing to work.

Reply:
It's maybe even possible to make that guess be a configurable parameter to the base `DatasetReader` class, similar to how the sorting keys are optional parameters to the bucket sampler, and guessed if they aren't provided.  It's trickier in this case because specifying the guess is hard, if there are lists.

Reply:
I guess, part of my thinking is, if we are going to require people to implement a new method, why don't we just go farther and fix the `TextField` API to remove those arguments in the first place?  If we aren't willing to do that (and there are very good reasons not to), then we should try hard to make it so that no one has to modify their code in order to use this new functionality.

Reply:
I'm still thinking about how to implement this efficiently, but a couple of thoughts:

> Yeah, but the difference between train and test will typically not be in `TextField`s.

Seq2seq models are one example where this isn't true.

> If we aren't willing to [change the `TextField` API] (and there are very good reasons not to), then we should try hard to make it so that no one has to modify their code in order to use this new functionality.

By "new functionality" are you referring to multi-process loading? Because they will already need to modify their `_read()` method to handle the sharding logic. Additionally overriding `apply_token_indexers` doesn't seem like a big ask.

Reply:
Hmm, ok, fair enough.  Still, though, you've given a helper method that makes the modification to `_read()` into a one-line change, unless I'm missing something.  The barrier to implementing this `apply_token_indexers` method feels a lot higher.

Also, having a default implementation that covers almost all use cases and an example of overriding it seems like a strict improvement over having no default implementation.

This is a pretty minor point for me, too, in the context of this PR.  Just do what you think is best.

Reply:
Yea, true. Well, I'll put some more thought into it and come back to it tomorrow. Like, my main concern is that we could be making assumptions which, when wrong, would introduce very subtle bugs.